rllib/connectors/agent.py

```diff
@@ -0,0 +1,501 @@
+            (obs, rewards, dones, infos, training_episode_infos),
+            ("obs", "rewards", "dones", "infos", "training_episode_infos"),
+        ):
+            assert isinstance(var, dict), (
+                f"EnvToPerAgentDataConnector expects {name} "
+                + "to be a MultiAgentDict."
+            )
+
+        env_id = ac_data.env_id
+        per_agent_data = []
+        for agent_id, obs in obs.items():
+            input_dict = {
+                SampleBatch.ENV_ID: env_id,
+                SampleBatch.REWARDS: rewards[agent_id],
+                # SampleBatch.DONES may be overridden by data from
+                # training_episode_infos next.
+                SampleBatch.DONES: dones[agent_id],
+                SampleBatch.NEXT_OBS: obs,
+            }
+            if SampleBatch.INFOS in self._view_requirements:
+                input_dict[SampleBatch.INFOS] = infos[agent_id]
+            if agent_id in training_episode_infos:
+                input_dict.update(training_episode_infos[agent_id])
+
+            per_agent_data.append(AgentConnectorDataType(env_id, agent_id, input_dict))
+
+        return per_agent_data
+
+    def to_config(self):
+        return EnvToPerAgentDataConnector.__name__, None
+
+    @staticmethod
+    def from_config(ctx: ConnectorContext, params: List[Any]):
+        return EnvToPerAgentDataConnector(ctx)
+
+
+register_connector(EnvToPerAgentDataConnector.__name__, EnvToPerAgentDataConnector)
+
+
+# Bridging between current obs preprocessors and connector.
+# We should not introduce any new preprocessors.
+# TODO(jungong) : migrate and implement preprocessor library in Connector framework.
+@DeveloperAPI
+class ObsPreprocessorConnector(AgentConnector):
+    """A connector that wraps around existing RLlib observation preprocessors.
+
+    This includes:
+    - OneHotPreprocessor for Discrete and Multi-Discrete spaces.
+    - GenericPixelPreprocessor and AtariRamPreprocessor for Atari spaces.
+    - TupleFlatteningPreprocessor and DictFlatteningPreprocessor for flattening
+      arbitrary nested input observations.
+    - RepeatedValuesPreprocessor for padding observations from RLlib Repeated
+      observation space.
+    """
+
+    def __init__(self, ctx: ConnectorContext):
+        super().__init__(ctx)
+
+        self._preprocessor = get_preprocessor(ctx.observation_space)(
+            ctx.observation_space, ctx.config.get("model", {})
+        )
+
+    def __call__(self, ac_data: AgentConnectorDataType) -> List[AgentConnectorDataType]:
+        d = ac_data.data
+        assert (
+            type(d) == dict
+        ), "Single agent data must be of type Dict[str, TensorStructType]"
+
+        if SampleBatch.OBS in d:
+            d[SampleBatch.OBS] = self._preprocessor.transform(d[SampleBatch.OBS])
+        if SampleBatch.NEXT_OBS in d:
+            d[SampleBatch.NEXT_OBS] = self._preprocessor.transform(
+                d[SampleBatch.NEXT_OBS]
+            )
+
+        return [ac_data]
```

Comment:
Sorry, another stupid question: Why return a list here?

Reply:
the API is now Any in, and List[Any] out.
the reason I want to use this API is so I can replace everything between env and policy with connector.
one of the important steps is to emit data for each agent individually from the multi-agent batch you get from the env.

Reply:
also, this allows someone to emit an empty [], meaning that the connector is not producing any data for this input.
useful for things like FrameSkippingAgentConnector.

Reply:
Would it ever be useful to have list-in + list-out?

Reply:
yeah, very good question.
I think it may be useful, but now that we have a stage where a connector can see all the multi-agent outputs from a single env, that need may not be as significant.
there may be a case where you would want to emit individual obs, then somehow batch some of them again. but I don't know any concrete use case yet.
for now, someone can achieve that by using a state-buffering kind of agent connector, and we can re-assess before 2.0 I guess.